simpl/middleware/context.py

```diff
@@ -0,0 +1,90 @@
+# Copyright (c) 2011-2015 Rackspace US, Inc.
+# All Rights Reserved.
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+"""Context WSGI Middleware.
+
+Creates a context for the WSGI call and adds the following to the context:
+- transaction_id: a UUID to identify the call (this can be passed in the
+        context to remote workers)
+- base_url: the URL of the incoming call (overrideable)
+
+The transaction id is returned in responses as an X-Transaction-Id header.
+
+Example usage:
+
+    # Disable all CORS checks:
+    import bottle
+    from simpl.middleware import context
+
+    app = bottle.default_app()
+    chain = context.ContextMiddleware(app, override_url="https://my_app.io")
+    bottle.run(app=chain)
+"""
+
+import logging
+import uuid
+
+from simpl import threadlocal
+
+LOG = logging.getLogger(__name__)
+
+
+class ContextMiddleware(object):  # pylint: disable=R0903
+
+    """Adds a call context to the call environ which holds call data."""
+
+    def __init__(self, app, override_url=None):
+        """Add a call context to the call environ which holds authn+z data."""
+        self.app = app
+        self.override_url = override_url
+
+    def __call__(self, environ, start_response):
+        """Handle WSGI Request."""
+        if self.override_url:
+            url = self.override_url
+        else:
+            # PEP333: wsgi.url_scheme, HTTP_HOST, SERVER_NAME, and SERVER_PORT
+            # can be used to reconstruct a request's complete URL
+            url = environ['wsgi.url_scheme'] + '://'
+            if environ.get('HTTP_HOST'):
+                url += environ['HTTP_HOST']
+            else:
+                url += environ['SERVER_NAME']
+
+                if environ['wsgi.url_scheme'] == 'https':
+                    if environ['SERVER_PORT'] != '443':
+                        url += ':' + environ['SERVER_PORT']
+                else:
+                    if environ['SERVER_PORT'] != '80':
+                        url += ':' + environ['SERVER_PORT']
```

Comment:
Just an FYI on how bottle does this, it might be more robust?

https://github.com/bottlepy/bottle/blob/master/bottle.py#L1329-L1340

Then accesses the rebuilt url with `geturl()`

https://github.com/bottlepy/bottle/blob/master/bottle.py#L1321